proposals/VK_EXT_map_memory_placed.asciidoc

```diff
@@ -0,0 +1,143 @@
+emulation environments when the application under emulation is 32-bit but
+the host userspace is 64-bit.  In order to avoid additional copies or
+address translation, the emulator can choose an address in the lower 32
+bits of the address space and request that the map be placed there.
+
+This functionality is already supported by the standard UNIX `mmap()`
+system call which provides an address hint flag as its first parameter as
+well as a `MAP_FIXED` flag which causes the `mmap()` to fail if it cannot
+place the map in exactly that location.  This extension proposes to add
+similar functionality to `vkMapMemory()`.
+
+
+## Solution Space
+
+It is possible to emulate this functionality using
+VK_EXT_external_memory_host by creating a placed memory mapping in some
+other way (such as a Linux memfd combined with conventional `mmap()`) and
+then importing the resulting map via VK_EXT_external_memory_host.  However,
+this requires every mappable Vulkan memory allocation to be a host pointer
+import which places restrictions available memory types and heaps and may
+have significant performance costs.
+
+One alternative approach to the emulation use-case would be to add a
+`VK_MEMORY_MAP_32BIT_ADDRESS_BIT_EXT` which does exactly what emulation
+wants and places the map in the lower 32 bits.  This is equivalent to the
+Linux `MAP_32BIT` flag.  However, this is less flexible and may be
+difficult to implement on Arm platforms where Linux `mmap()` does not
+support `MAP_32BIT` since drivers are likely relying on `mmap()` to
+implement `vkMapMemory()`.
+
+## Proposal
+
+### API Features
+
+The following features are exposed by this extension:
+
+[source,c]
+----
+typedef enum VkMemoryMapFlagBits {
+    VK_MEMORY_MAP_PLACED_BIT_EXT = 0x00000001,
+} VkMemoryMapFlagBits;
+
+typedef enum VkMemoryUnmapFlagBits {
+    VK_MEMORY_UNMAP_RESERVE_BIT_EXT = 0x00000001,
+} VkMemoryMapFlagBits;
+----
+
+ . `memoryMapPlaced` indicates support for placed memory maps
+ . `memoryMapRangePlaced` indicates support for placed memory maps of
+   subranges of the `VkDeviceMemory` object.
+ . `memoryUnmapReserve` indicates support for
+   `VK_MEMORY_UNMAP_RESERVE_BIT_EXT`.
+
+When `VK_MEMORY_MAP_PLACED_BIT_EXT` is set in the `flags` member of
+`VkMemoryMapInfoEXT` and a `VkMapMemoryPlacedInfoEXT` structure is included
+in the `pNext` chain of `VkMemoryMapInfoEXT`, the implementation attempts
+to place the map at the address specified by
+`VkMapMemoryPlacedInfoEXT::pPlacedAddress`.
+If the map cannot be placed exactly at the specified virtual address,
+`vkMapMemory2EXT()` returns `VK_ERROR_MEMORY_MAP_FAILED`.
+
+When `VK_MEMORY_UNMAP_RESERVE_BIT_EXT` is set in the `flags` member of
+`VkMemoryUnmapInfoEXT`, the address range currently occupied by the memory
+map will be atomically re-reserved by `vkUnmapMemory()` so that another
+call to `mmap()` or `vkMapMemory()` will not re-use that address range
+without an address hint or `VK_MEMORY_MAP_PLACED_BIT_EXT`, respectively.
```

Comment:
Yeah, spec'ing this is tricky.  "remains reserved" is ok, but the fact that it's atomic is necessary.  If all we cared about was the ability to re-reserve, the client could do the reservation with `mmap()` after it returns.  We need to ensure that nothing can get in between to solve the relevant race or there's no point in having this part of the Vulkan extension.  See https://github.com/KhronosGroup/Vulkan-Docs/issues/1832#issuecomment-1199820012 for more details on why this feature is needed.

Reply:
Yea but you only need to talk about "atomic re-reserve" if you imply that unmap unreserves the range instead of just unmapping it. So you can shortcut that by saying it's never unreserved in the first place.

Reply:
Possibly?  I'm already waving my hands a lot by talking about "reserved" VA ranges when we have no such concept in the spec and it's really an OS concept that's entirely out of the Vulkan's jurisdiction.  Let me think on it.

Reply:
How about ", the memory object will be unmapped but the address range will remain reserved so that another..."?  I still feel like we need to say somewhere explicitly that it's all atomic and another `mmap()` in another thread can't race with it and steal your VA range.

Reply:
Yea that works.

Reply:
Done.